frontend/src/metabase/query_builder/components/NativeQueryEditor/CodeMirrorEditor/completers.tsx

```diff
@@ -0,0 +1,252 @@
+      if (!word) {
+        return null;
+      }
+
+      const { data } = await getAutocompleteSuggestions(
+        { databaseId, matchStyle, query: word.text.trim() },
+        false,
+      );
+      if (!data) {
+        return null;
+      }
+
+      return {
+        from: word.from,
+        options: data.map(([value, meta]) => ({
+          label: value,
+          detail: meta,
+          boost: 50,
+        })),
+        validFor(text: string) {
+          if (data.length >= AUTOCOMPLETE_SUGGESTIONS_LIMIT) {
+            // If there are more suggestions than the limit, we want subsequent
+            // edits to fetch more completions because we can't assume that we've seen all
+            // suggestions
+            return false;
+          }
+
+          return text.startsWith(word.text);
+        },
+      };
+    },
+    [databaseId, matchStyle, getAutocompleteSuggestions],
+  );
+}
+
+// Completes snippet names when inside a snippet tag
+export function useSnippetCompletion() {
+  const { data: snippets } = useListSnippetsQuery();
+
+  return useCallback(
+    function completeSnippetTags(context: CompletionContext) {
+      const tag = matchTagAtCursor(context.state, {
+        allowOpenEnded: true,
+        position: context.pos,
+      });
+
+      if (tag?.type !== "snippet") {
+        return null;
+      }
+
+      const query = tag.content.text;
+      const results =
+        snippets?.filter(snippet =>
+          snippet.name.toLowerCase().includes(query.toLowerCase()),
+        ) ?? [];
+
+      return {
+        from: tag.content.from,
+        validFor(text: string) {
+          return text.startsWith(query);
+        },
+        options: results.map(snippet => ({
+          label: snippet.name,
+          apply: tag.hasClosingTag ? snippet.name : `${snippet.name} }}`,
+          detail: t`Snippet`,
+          boost: 50,
+        })),
+      };
+    },
+    [snippets],
+  );
+}
+
+// Keep this in sync with the limit in the backend code at
+// `autocomplete-cards` in src/metabase/api/database.clj
+const AUTOCOMPLETE_CARD_SUGGESTIONS_LIMIT = 50;
+
+type CardTagCompletionOptions = {
+  databaseId?: DatabaseId | null;
+};
+
+// Completes card names when inside a card tag
+export function useCardTagCompletion({ databaseId }: CardTagCompletionOptions) {
+  const [getCardAutocompleteSuggestions] =
+    useLazyGetCardAutocompleteSuggestionsQuery();
+  return useCallback(
+    async function completeCardTags(context: CompletionContext) {
+      if (databaseId == null) {
+        return null;
+      }
+
+      const tag = matchTagAtCursor(context.state, {
+        allowOpenEnded: true,
+        position: context.pos,
+      });
+      if (tag?.type !== "card") {
+        return null;
+      }
+
+      const query = tag.content.text.replace(/^#/, "").trim();
+      if (!query) {
+        return null;
+      }
+
+      const { data } = await getCardAutocompleteSuggestions({
+        databaseId,
+        query,
+      });
```

Comment:
I'm don't think we want that since we don't have a way to invalidate that cache which might lead to confusing results. The autocomplete endpoint has a (short-lived) `Cache-Control` header so we're using the browsers' request cache instead.

I'm happy to change this in the future, but I think this is out of scope for this PR.